### PR TITLE
Gp2 1576 casestudy csv

### PR DESCRIPTION
**Pull request by webbyfox**
This changeset adds a download button in admin UI to export all case studies attributes into CSV.

### Workflow

- [x] Ticket exists in Jira https://uktrade.atlassian.net/browse/1576
- [x] Jira ticket has the correct status.
- [x] [Changelog](CHANGELOG.md) entry added.
- [x] This PR can be merged by reviewers. (If unticked, please leave for the author to merge)

On the case study admin index page, we have download csv button 

image

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -32,15 +36,6 @@
 - GP2-1735 - update lessons
 - GP2-1632-generic-ep-pages
 - GP2-1782 - ep section progress
-
```
Comment (stevejalim) [inline CHANGELOG.md:35]: Thanks for cleaning up this 💅

File: cms_extras/modeladmin.py [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
 from wagtail.contrib.modeladmin.options import ModelAdmin, modeladmin_register
+from wagtail.contrib.modeladmin.views import IndexView
 
 from core.models import CaseStudy
 
 
 class CaseStudyAdminButtonHelper(ButtonHelper):
-
-    view_button_classnames = ['button-small', 'icon', 'icon-doc']
+    view_button_classnames = ['button-small', 'icon', 'icon-doc', 'btn-group']
```
Comment (stevejalim) [inline cms_extras/modeladmin.py:13]: Just curious - what's this `btn-group` for?

File: cms_extras/modeladmin.py [hunk 3]
```diff
@@ -26,12 +29,97 @@ def get_buttons_for_obj(self, obj, exclude=None, classnames_add=None, classnames
...
+                for field in self.list_export
+                if field not in self.columns_to_convert
+            }
+        )
+
+        for item in queryset:
+            yield self.write_csv_row(writer, self.to_row_dict(item))
+
+    def create_row_by_tag(self, row_dict, processed_row):
+        processed_row_list = []
```
Comment (stevejalim) [inline cms_extras/modeladmin.py:61]: I'd add a comment here explaining that this expands the stub of a single processed row into its various permutations, just to help our future selves

File: cms_extras/modeladmin.py [hunk 4]
```diff
@@ -26,12 +29,97 @@ def get_buttons_for_obj(self, obj, exclude=None, classnames_add=None, classnames
...
+        for f in self.columns_to_convert:
+            # No tagged value then just append row for write
+            attr_values = [] if row_dict.get(f) == '-' else row_dict.get(f)
```
Comment (stevejalim) [inline cms_extras/modeladmin.py:64]: tiny thing - is it worth moving this to be the 'else' of an 'if / elif / else  ' below to avoid wasted computation if `f` is `'get_related_pages'` or `'modified'`?

File: cms_extras/modeladmin.py [hunk 5]
```diff
@@ -26,12 +29,97 @@ def get_buttons_for_obj(self, obj, exclude=None, classnames_add=None, classnames
...
+
+    def write_csv_row(self, writer, row_dict):
+        processed_row = {}
+        common_fields = {k: v for k, v in row_dict.items() if k in ['title', 'summary_context', 'lead_title']}
+        for field, value in common_fields.items():
+            preprocess_function = self.get_preprocess_function(field, value, self.FORMAT_CSV)
+            processed_value = preprocess_function(value) if preprocess_function else value
+            processed_row[field] = processed_value
+        return self.write_multiple_rows(writer, self.create_row_by_tag(row_dict, processed_row))
+
```
Comment (stevejalim) [inline cms_extras/modeladmin.py:95]: Not a change request but this whole group of methods feels like it'd benefit from type hinting on the signatures - what do you think?

File: cms_extras/modeladmin.py [hunk 6]
```diff
@@ -81,5 +169,11 @@ def get_related_pages(self, obj):
 
     get_related_pages.short_description = 'Associated pages'
 
+    def association(self, obj):
+        pass
+
+    def attribute(self, obj):
+        pass
+
```
Comment (stevejalim) [inline cms_extras/modeladmin.py:177]: Why are these here if they don't (I think?) render anything in the modeladmin?

File: config/settings.py [hunk 7]
```diff
@@ -9,6 +9,7 @@
 from .utils import get_wagtail_transfer_configuration
 
 ROOT_DIR = environ.Path(__file__) - 2
+APPS_DIR = ROOT_DIR.path('core')
```
Comment (stevejalim) [inline config/settings.py:12]: I think this could be better named as `CORE_APP_DIR` because `APPS_DIR` sounds like it's all the apps not just `core`

File: config/settings.py [hunk 8]
```diff
@@ -95,9 +96,15 @@
...
-        'APP_DIRS': True,
+        'DIRS': [
+            APPS_DIR.path('templates'),
+        ],
         'OPTIONS': {
             'debug': True,
+            'loaders': [
+                'django.template.loaders.filesystem.Loader',
+                'django.template.loaders.app_directories.Loader',
+            ],
```
Comment (stevejalim) [inline config/settings.py:107]: Am surprised we needed to add these - is this specifically to get the CSV export to work?

File: core/templates/modeladmin/index.html [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                    <div class="right">
+
+                        {% if user_can_create %}
+                            <div class="actionbutton">
+                                {% include 'modeladmin/includes/button.html' with button=view.button_helper.add_button %}
+
+
+                        {% if view.list_export %}
+                            <div class="dropdown dropdown-button match-width col">
+                                <a href="?export=csv&{{ request.GET.urlencode }}" class="button bicolor button--icon">{% icon name="download" wrapped=1 %}{% trans 'Download CSV' %}</a>
```
Comment (stevejalim) [inline core/templates/modeladmin/index.html:36]: Is this the only line that's customised? If so, It'd be better to only override `modeladmin/index.html` for `{% block header_extra %}` and reduce the duplication in this template. 

If there are more changes in this template, would be cool to make them easy to find via comments, in case someone overwrites them if, say, the `wagtail.contrib.modeladmin` is refactored

**Review by stevejalim — COMMENTED** [review]
Looking pretty great - nice one! ⭐ 

Added a bunch of small questions and two comments that might need attention (the ones about `APPS_DIR` and `modeladmin/index.html`

Let me know what you think and am happy to approve after that.

File: cms_extras/modeladmin.py (hunk 6 above)
Comment (webbyfox) [reply cms_extras/modeladmin.py:186]: Added comment in code for this

File: core/templates/modeladmin/index.html (hunk 9 above)
Comment (webbyfox) [reply core/templates/modeladmin/index.html]: Thanks for picking this. Good catch

**Review by stevejalim — APPROVED** [review]
LGTM - thanks!

I think we could put _more_ type hinting on some signatures, but it's not a blocker at all 🚢
